iothub/service/src/Authentication/IotHubConnectionStringParser.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Azure.Devices
+{
+    internal sealed class IotHubConnectionStringParser
+    {
+        internal static IotHubConnectionString Parse(string iotHubConnectionString)
+        {
+            IDictionary<string, string> map = ToDictionary(iotHubConnectionString, IotHubConnectionStringConstants.ValuePairDelimiter, IotHubConnectionStringConstants.ValuePairSeparator);
+
+            string hostName = GetConnectionStringValue(map, IotHubConnectionStringConstants.HostNamePropertyName);
+            string sharedAccessKeyName = GetConnectionStringOptionalValue(map, IotHubConnectionStringConstants.SharedAccessKeyNamePropertyName);
+            string sharedAccessKey = GetConnectionStringOptionalValue(map, IotHubConnectionStringConstants.SharedAccessKeyPropertyName);
+            string sharedAccessSignature = GetConnectionStringOptionalValue(map, IotHubConnectionStringConstants.SharedAccessSignaturePropertyName);
+
+            Validate(hostName, sharedAccessKeyName, sharedAccessKey, sharedAccessSignature);
+
+            return new IotHubConnectionString(hostName, sharedAccessKeyName, sharedAccessKey, sharedAccessSignature);
+        }
+
+        private static void Validate(string hostName, string sharedAccessKeyName, string sharedAccessKey, string sharedAccessSignature)
+        {
+            string iotHubName = GetIotHubName(hostName);
+            if (string.IsNullOrWhiteSpace(sharedAccessKeyName))
+            {
+                throw new ArgumentException("Should specify either SharedAccessKeyName.");
```

Comment:
Is the shared access key name mandatory? The message there seems weird - either shared access key name or...?

Reply:
This message was to validate either SharedAccessKeyName or deviceId is always present. 
Updated the message.

Reply:
Shared access key name is mandatory. I have checked it in Java world (service side) and also verified that it is required when creating SAS tokens.

Reply:
Confirmed it with Tim that we dont actually need to validate presence of shared access key name.
Removed the validation check.